<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Integrative pipeline for cell type deconvolution from bulk RNAseq using first and second generation methods • multideconv</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Integrative pipeline for cell type deconvolution from bulk RNAseq using first and second generation methods">
<meta name="description" content="R package that reduces dimensionality and eliminates redundancy through unsupervised filtering and iterative correlation analyses. Built on top of existing frameworks like immunedeconv (Sturm et al. 2019) and omnideconv (Dietrich et al. 2024), multideconv harmonizes outputs across methods to identify robust cell subgroups and mitigate signature-driven heterogeneity.">
<meta property="og:description" content="R package that reduces dimensionality and eliminates redundancy through unsupervised filtering and iterative correlation analyses. Built on top of existing frameworks like immunedeconv (Sturm et al. 2019) and omnideconv (Dietrich et al. 2024), multideconv harmonizes outputs across methods to identify robust cell subgroups and mitigate signature-driven heterogeneity.">
<meta property="og:image" content="https://mhurtado13.github.io/multideconv/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">multideconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/multideconv.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/VeraPancaldiLab/multideconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="multideconv">multideconv<a class="anchor" aria-label="anchor" href="#multideconv"></a>
</h1>
</div>
<p>An integrative pipeline for combining first and second generation cell type deconvolution results</p>
<!-- badges: start -->
<!-- badges: end -->
<p align="center">
<img src="reference/figures/overview.png?raw=true"></p>
<p align="center">
<em>Figure 1. A schematic overview of the <code>multideconv</code> pipeline</em>
</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To avoid GitHub API rate limit issues during installation, we recommend setting up GitHub authentication by creating and storing a Personal Access Token (PAT). You can do this with the following steps:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages(c("usethis", "gitcreds"))</span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/github-token.html" class="external-link">create_github_token</a></span><span class="op">(</span><span class="op">)</span> <span class="co">#Create a Personal Access Token (if you don't have)</span></span>
<span><span class="fu">gitcreds</span><span class="fu">::</span><span class="fu"><a href="https://gitcreds.r-lib.org/reference/gitcreds_get.html" class="external-link">gitcreds_set</a></span><span class="op">(</span><span class="op">)</span> <span class="co">#Add the token</span></span></code></pre></div>
<p>You can install the development version of <code>multideconv</code> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("pak")</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"VeraPancaldiLab/multideconv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="general-usage">General usage<a class="anchor" aria-label="anchor" href="#general-usage"></a>
</h2>
<p>These are basic examples which shows you how to use <code>multideconv</code> for different tasks. For a detailed tutorial, see <a href="https://VeraPancaldiLab.github.io/multideconv/articles/multideconv.html" class="external-link">Get started</a></p>
<p>Before running <code>multideconv</code>, make sure to set your working directory. The <code>Results/</code> folder, where outputs will be saved, will be created in this directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">'~/path/to/directory'</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/VeraPancaldiLab/multideconv" class="external-link">multideconv</a></span><span class="op">)</span></span></code></pre></div>
<p>The function calculates cell abundance based on cell type signatures using different methods and signatures through the function <code>compute.deconvolution</code> which takes as input the bulk RNAseq gene expression matrix either as raw or normalized counts.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="reference/compute.deconvolution.html">compute.deconvolution</a></span><span class="op">(</span><span class="va">raw.counts</span>, normalized <span class="op">=</span> <span class="cn">T</span>, credentials.mail <span class="op">=</span> <span class="st">"xxxx"</span>, credentials.token <span class="op">=</span> <span class="st">"xxxxxx"</span>, file_name <span class="op">=</span> <span class="st">"Tutorial"</span><span class="op">)</span> </span>
<span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="reference/compute.deconvolution.html">compute.deconvolution</a></span><span class="op">(</span><span class="va">raw.counts</span>, normalized <span class="op">=</span> <span class="cn">T</span>, credentials.mail <span class="op">=</span> <span class="st">"xxxx"</span>, credentials.token <span class="op">=</span> <span class="st">"xxxxxx"</span>, methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quantiseq"</span>, <span class="st">"MCP"</span>, <span class="st">"XCell"</span>, <span class="st">"DWLS"</span><span class="op">)</span>, file_name <span class="op">=</span> <span class="st">"Test"</span><span class="op">)</span> </span>
<span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="reference/compute.deconvolution.html">compute.deconvolution</a></span><span class="op">(</span><span class="va">raw.counts</span>, normalized <span class="op">=</span> <span class="cn">T</span>, credentials.mail <span class="op">=</span> <span class="st">"xxxx"</span>, credentials.token <span class="op">=</span> <span class="st">"xxxxxx"</span>, signatures_exclude <span class="op">=</span> <span class="st">"BPRNACan"</span>, file_name <span class="op">=</span> <span class="st">"Tutorial"</span><span class="op">)</span></span>
<span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="reference/compute.deconvolution.html">compute.deconvolution</a></span><span class="op">(</span><span class="va">raw.counts</span>, normalized <span class="op">=</span> <span class="cn">T</span>, credentials.mail <span class="op">=</span> <span class="st">"xxxx"</span>, credentials.token <span class="op">=</span> <span class="st">"xxxxxx"</span>, sc_deconv <span class="op">=</span> <span class="cn">T</span>, sc_matrix <span class="op">=</span> <span class="va">sc.object</span>, cell_label <span class="op">=</span> <span class="va">cell_labels</span>, sample_label <span class="op">=</span> <span class="va">bath_ids</span>, name_sc_signature <span class="op">=</span> <span class="st">"Signature_test"</span>, file_name <span class="op">=</span> <span class="st">"Test"</span><span class="op">)</span></span></code></pre></div>
<p><strong>NOTE</strong>: <code>CIBERSORTx</code> is included in the deconvolution methods, but it’s not an open-source program. To run it, please ask for a token in <a href="https://cibersortx.stanford.edu/register.php" class="external-link">CIBERSORTx</a> and once obtained, provided your username and password on the parameters <code>credentials.mail</code> and <code>credentials.token</code>.</p>
<p>Also CIBERSORTx ask to have <a href="https://kinsta.com/blog/install-docker-ubuntu/#installing-docker-desktop-on-ubuntu" class="external-link">docker</a> install in your computer. Once installed, be sure to concede permission to manage docker as a non-root user. For verify this, go to the terminal and run:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>docker ps</span></code></pre></div>
<p>If you don’t have any error, congrats you are go to go!</p>
<p>If you receive an error of permission, just run:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>sudo groupadd docker</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>sudo usermod <span class="sc">-</span>aG docker <span class="sc">$</span>{USER}</span></code></pre></div>
<p>Then restart your computer and try again</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>docker ps</span></code></pre></div>
<p>Now, you should be able to run deconvolution using CIBERSORTx without any problems :)</p>
<p>For processing the deconvolution features obtained from <code>compute.deconvolution</code>, you can use the <code>compute.deconvolution.analysis</code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">processed_deconvolution</span> <span class="op">=</span> <span class="fu"><a href="reference/compute.deconvolution.analysis.html">compute.deconvolution.analysis</a></span><span class="op">(</span><span class="va">deconvolution</span>, corr <span class="op">=</span> <span class="fl">0.7</span>, seed <span class="op">=</span> <span class="fl">123</span>, return <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>Users can also compute second-generation deconvolution methods using their single cell data. For this use the function <code>compute_sc_deconvolution_methods</code>. Remember that this function is already included in <code>compute.deconvolution</code> when setting <code>sc_deconv = T</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconv_sc</span> <span class="op">=</span> <span class="fu"><a href="reference/compute_sc_deconvolution_methods.html">compute_sc_deconvolution_methods</a></span><span class="op">(</span><span class="va">raw_counts</span>, <span class="va">sc_object</span>, <span class="va">sc_metadata</span>, <span class="va">cell_annotations</span>, <span class="va">samples_ids</span>, <span class="va">name_object</span>, normalized <span class="op">=</span> <span class="cn">T</span>, n_cores <span class="op">=</span> <span class="fl">4</span>, cbsx_name <span class="op">=</span> <span class="st">"XXX"</span>, cbsx_token <span class="op">=</span> <span class="st">"XXX"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cell-types-nomenclature">Cell types nomenclature<a class="anchor" aria-label="anchor" href="#cell-types-nomenclature"></a>
</h2>
<p><code>multideconv</code> works based on established cell naming conventions (Figure 2) to simplify analysis and processing. Thus, if you would like to use your own deconvolution results or signatures, please make sure to follow these formats.</p>
<p align="center">
<img src="reference/figures/cell_types.png?raw=true"></p>
<p align="center">
<em>Figure 2. Cell types nomenclature for <code>multideconv</code></em>
</p>
</div>
<div class="section level2">
<h2 id="how-to-add-cell-types-other-than-the-ones-present-in-the-nomenclature">How to add cell types other than the ones present in the nomenclature?<a class="anchor" aria-label="anchor" href="#how-to-add-cell-types-other-than-the-ones-present-in-the-nomenclature"></a>
</h2>
<p>If you want <code>multideconv</code> to consider other cells, it is pretty simple! Just use the argument cells_extra in the function <code><a href="reference/compute.deconvolution.analysis.html">compute.deconvolution.analysis()</a></code>.</p>
<p>Let’s say you want to add mesenchymal and basophils cells:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">processed_deconvolution</span> <span class="op">=</span> <span class="fu"><a href="reference/compute.deconvolution.analysis.html">compute.deconvolution.analysis</a></span><span class="op">(</span><span class="va">deconvolution</span>, corr <span class="op">=</span> <span class="fl">0.7</span>, seed <span class="op">=</span> <span class="fl">123</span>, cells_extra <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mesenchymal"</span>, <span class="st">"basophils"</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>And that’s it, just make sure the name you are putting in cells_extra is exactly the name of your cells in your deconvolution matrix!</p>
</div>
<div class="section level2">
<h2 id="how-to-add-other-signatures">How to add other signatures?<a class="anchor" aria-label="anchor" href="#how-to-add-other-signatures"></a>
</h2>
<p>You can include other signatures into the analysis by adding them as .txt into the folder <code>Results/custom_signatures</code>.</p>
</div>
<div class="section level2">
<h2 id="how-does-my-single-cell-data-is-used-for-deconvolution">How does my single cell data is used for deconvolution?<a class="anchor" aria-label="anchor" href="#how-does-my-single-cell-data-is-used-for-deconvolution"></a>
</h2>
<p><code>multideconv</code> includes second generation methods that allows users to input single cell data and use it for constructing signatures ‘on-the-fly’ and deconvolve the bulk RNAseq data.</p>
<p>Because most of the times scRNAseq data can be big and sparse matrices, we applied ‘preprocessing’ steps to avoid crashing the computer. For this, <code>multideconv</code> construct metacells per cell type and patient from the single cell data using the KNN algorithm. This is done essentially by executing different tasks per cluster in parallel for N workers and once finished, it computes deconvolution using the reduced single cell object.</p>
<p><strong>NOTE</strong>: If you encounter missing packages for certain deconvolution methods (e.g., <code>MuSiC</code>, <code>SCDC</code>) during the pipeline, make sure to install the versions maintained by the <code>omnideconv</code> team. These versions are patched to work seamlessly with <code>omnideconv</code> functions:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"omnideconv/MuSiC"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"omnideconv/SCDC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="issues">Issues<a class="anchor" aria-label="anchor" href="#issues"></a>
</h2>
<p>If you encounter any problems or have questions about the package, we encourage you to open an issue <a href="https://github.com/VeraPancaldiLab/multideconv/issues" class="external-link">here</a>. We’ll do our best to assist you!</p>
</div>
<div class="section level2">
<h2 id="authors">Authors<a class="anchor" aria-label="anchor" href="#authors"></a>
</h2>
<p><code>multideconv</code> was developed by <a href="https://github.com/mhurtado13" class="external-link">Marcelo Hurtado</a> in supervision of <a href="https://github.com/VeraPancaldi" class="external-link">Vera Pancaldi</a> and is part of the <a href="https://github.com/VeraPancaldiLab" class="external-link">Pancaldi</a> team. Currently, Marcelo is the primary maintainer of this package.</p>
</div>
<div class="section level2">
<h2 id="citing-multideconv">Citing <code>multideconv</code>
<a class="anchor" aria-label="anchor" href="#citing-multideconv"></a>
</h2>
<p>If you use <code>multideconv</code> in a scientific publication, we would appreciate citation to the :</p>
<p><em>Hurtado, M., Essabbar, A., Khajavi, L., &amp; Pancaldi, V. (2025). multideconv – Integrative pipeline for cell type deconvolution from bulk RNAseq using first and second generation methods. bioRxiv. <a href="https://doi.org/10.1101/2025.04.29.651220" class="external-link uri">https://doi.org/10.1101/2025.04.29.651220</a></em></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/VeraPancaldiLab/multideconv/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3) + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing multideconv</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Marcelo Hurtado <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0004-6712-0864" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marcelo Hurtado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
