<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute deconvolution — compute.deconvolution • multideconv</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute deconvolution — compute.deconvolution"><meta name="description" content="The function calculates cell abundance based on cell type signatures using different methods and signatures. Methods available are Quantiseq, MCP, XCell, CIBERSORTx, EpiDISH, DWLS and DeconRNASeq. Provided signatures included signatures based on bulk and methylation data (7 methods and 10 signature in total). Signatures are present in the src/signatures directory, user can add its own signatures by adding the .txt files in this same folder. Second generation methods to perform deconvolution based on single cell data are also available if scRNAseq object is provided."><meta property="og:description" content="The function calculates cell abundance based on cell type signatures using different methods and signatures. Methods available are Quantiseq, MCP, XCell, CIBERSORTx, EpiDISH, DWLS and DeconRNASeq. Provided signatures included signatures based on bulk and methylation data (7 methods and 10 signature in total). Signatures are present in the src/signatures directory, user can add its own signatures by adding the .txt files in this same folder. Second generation methods to perform deconvolution based on single cell data are also available if scRNAseq object is provided."><meta property="og:image" content="https://mhurtado13.github.io/multideconv/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">multideconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/multideconv.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/VeraPancaldiLab/multideconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute deconvolution</h1>
      <small class="dont-index">Source: <a href="https://github.com/VeraPancaldiLab/multideconv/blob/main/R/cell_deconvolution.R" class="external-link"><code>R/cell_deconvolution.R</code></a></small>
      <div class="d-none name"><code>compute.deconvolution.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function calculates cell abundance based on cell type signatures using different methods and signatures. Methods available are Quantiseq, MCP, XCell, CIBERSORTx, EpiDISH, DWLS and DeconRNASeq. Provided signatures included signatures based on bulk and methylation data (7 methods and 10 signature in total). Signatures are present in the src/signatures directory, user can add its own signatures by adding the .txt files in this same folder. Second generation methods to perform deconvolution based on single cell data are also available if scRNAseq object is provided.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compute.deconvolution</span><span class="op">(</span></span>
<span>  <span class="va">raw.counts</span>,</span>
<span>  methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quantiseq"</span>, <span class="st">"CBSX"</span>, <span class="st">"Epidish"</span>, <span class="st">"DeconRNASeq"</span>, <span class="st">"DWLS"</span><span class="op">)</span>,</span>
<span>  signatures_exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalized <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  doParallel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  workers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  create_signature <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  credentials.mail <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  credentials.token <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sc_deconv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sc_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sc_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  methods_sc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Autogenes"</span>, <span class="st">"BayesPrism"</span>, <span class="st">"Bisque"</span>, <span class="st">"CPM"</span>, <span class="st">"MuSic"</span>, <span class="st">"SCDC"</span><span class="op">)</span>,</span>
<span>  cell_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  methods_sig <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DWLS"</span>, <span class="st">"CIBERSORTx"</span>, <span class="st">"MOMF"</span>, <span class="st">"BSeqsc"</span><span class="op">)</span>,</span>
<span>  name_sc_signature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  file_name <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-raw-counts">raw.counts<a class="anchor" aria-label="anchor" href="#arg-raw-counts"></a></dt>
<dd><p>A matrix with the raw counts (samples as columns and genes symbols as rows)</p></dd>


<dt id="arg-methods">methods<a class="anchor" aria-label="anchor" href="#arg-methods"></a></dt>
<dd><p>A character vector with the deconvolution methods to run. Default are "Quantiseq", "MCP", "xCell", "CBSX", "Epidish", "DeconRNASeq", "DWLS"</p></dd>


<dt id="arg-signatures-exclude">signatures_exclude<a class="anchor" aria-label="anchor" href="#arg-signatures-exclude"></a></dt>
<dd><p>A character vector with the signatures to exclude from the src/signatures folder.</p></dd>


<dt id="arg-normalized">normalized<a class="anchor" aria-label="anchor" href="#arg-normalized"></a></dt>
<dd><p>If raw.counts are not available, user can input its normalized counts. In that case this argument need to be set to False.</p></dd>


<dt id="arg-doparallel">doParallel<a class="anchor" aria-label="anchor" href="#arg-doparallel"></a></dt>
<dd><p>Whether to do or not parallelization. Only CBSX and DWLS methods will run in parallel.</p></dd>


<dt id="arg-workers">workers<a class="anchor" aria-label="anchor" href="#arg-workers"></a></dt>
<dd><p>Number of processes available to run on parallel. If no number is set, this will correspond to detectCores() - 1</p></dd>


<dt id="arg-return">return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>Whether to save or not the csv file with the deconvolution features</p></dd>


<dt id="arg-create-signature">create_signature<a class="anchor" aria-label="anchor" href="#arg-create-signature"></a></dt>
<dd><p>Whether to create or not the signatures using the methods MOMF, CBSX, DWLS and BSeq-SC. If TRUE, sc_matrix shuld be provide.</p></dd>


<dt id="arg-credentials-mail">credentials.mail<a class="anchor" aria-label="anchor" href="#arg-credentials-mail"></a></dt>
<dd><p>(Optional) Credential email for running CIBERSORTx. If not provided, CIBERSORTx method will not be run.</p></dd>


<dt id="arg-credentials-token">credentials.token<a class="anchor" aria-label="anchor" href="#arg-credentials-token"></a></dt>
<dd><p>(Optional) Credential token for running CIBERSORTx. If not provided, CIBERSORTx method will not be run.</p></dd>


<dt id="arg-sc-deconv">sc_deconv<a class="anchor" aria-label="anchor" href="#arg-sc-deconv"></a></dt>
<dd><p>Whether to run or not deconvolution methods based on single cell.</p></dd>


<dt id="arg-sc-matrix">sc_matrix<a class="anchor" aria-label="anchor" href="#arg-sc-matrix"></a></dt>
<dd><p>If sc_deconv = T, the matrix of counts across cells from the scRNAseq object is provided.</p></dd>


<dt id="arg-sc-metadata">sc_metadata<a class="anchor" aria-label="anchor" href="#arg-sc-metadata"></a></dt>
<dd><p>Dataframe with metadata from the single cell object. The matrix should include the columns cell_label and sample_label.</p></dd>


<dt id="arg-methods-sc">methods_sc<a class="anchor" aria-label="anchor" href="#arg-methods-sc"></a></dt>
<dd><p>A character vector with the sc-deconvolution methods to run. Default are "Autogenes", "BayesPrism", "Bisque", "CPM", "MuSic", "SCDC"</p></dd>


<dt id="arg-cell-label">cell_label<a class="anchor" aria-label="anchor" href="#arg-cell-label"></a></dt>
<dd><p>If sc_deconv = T, a character vector indicating the cell labels (same order as the count matrix)</p></dd>


<dt id="arg-sample-label">sample_label<a class="anchor" aria-label="anchor" href="#arg-sample-label"></a></dt>
<dd><p>If sc_deconv = T, a character vector indicating the cell samples IDs (same order as the count matrix)</p></dd>


<dt id="arg-cell-markers">cell_markers<a class="anchor" aria-label="anchor" href="#arg-cell-markers"></a></dt>
<dd><p>Named list with the genes markers names as Symbol per cell types to be used to create the signature using the BSeq-SC method. If NULL, the method will be ignored during the signature creation.</p></dd>


<dt id="arg-methods-sig">methods_sig<a class="anchor" aria-label="anchor" href="#arg-methods-sig"></a></dt>
<dd><p>A character vector specifying which methods to run. Options are "DWLS", "CIBERSORTx", "MOMF", and "BSeqsc". Default runs all available methods.</p></dd>


<dt id="arg-name-sc-signature">name_sc_signature<a class="anchor" aria-label="anchor" href="#arg-name-sc-signature"></a></dt>
<dd><p>If sc_deconv = T, the name you want to give to the signature generated</p></dd>


<dt id="arg-file-name">file_name<a class="anchor" aria-label="anchor" href="#arg-file-name"></a></dt>
<dd><p>File name for the csv files and plots saved in the Results/ directory</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix of cell type deconvolution features across samples</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Sturm, G., Finotello, F., Petitprez, F., Zhang, J. D., Baumbach, J., Fridman, W. H., ..., List, M., Aneichyk, T. (2019). Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology.
Bioinformatics, 35(14), i436-i445. https://doi.org/10.1093/bioinformatics/btz363</p>
<p>Benchmarking second-generation methods for cell-type deconvolution of transcriptomic data. Dietrich, Alexander and Merotto, Lorenzo and Pelz, Konstantin and Eder, Bernhard and Zackl, Constantin and Reinisch, Katharina and
Edenhofer, Frank and Marini, Federico and Sturm, Gregor and List, Markus and Finotello, Francesca. (2024) https://doi.org/10.1101/2024.06.10.598226</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"raw_counts"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"cell_labels"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sample_labels"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"metacells_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"metacells_metadata"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pseudobulk"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">deconv</span> <span class="op">=</span> <span class="fu">compute.deconvolution</span><span class="op">(</span><span class="va">raw_counts</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                               methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Epidish"</span>, <span class="st">"DeconRNASeq"</span><span class="op">)</span>, return <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Performing TPM normalization ................................................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting input to matrix.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running deconvolution using the following methods...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * Epidish</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * DeconRNASeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The following method-signature combinations are going to be calculated...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Methods</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * Epidish</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * DeconRNASeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Signatures</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * BPRNACan</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * BPRNACan3DProMet</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * BPRNACanProMet</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * CBSX-HNSCC-scRNAseq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * CBSX-Melanoma-scRNAseq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * CBSX-NSCLC-PBMCs-scRNAseq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * CCLE-TIL10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * LM22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> * TIL10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: generics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘generics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     setequal, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, is.unsorted, lapply,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rank, rbind, rownames, sapply, saveRDS, table, tapply, unique,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘pcaMethods’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     loadings</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 10  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 10  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 10  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 10  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 8  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 6  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10    PC11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282 0.00234</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084 0.98318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 11  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10    PC11    PC12    PC13    PC14   PC15    PC16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282 0.00234 0.00183 0.00178 0.00162 0.0014 0.00123</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084 0.98318 0.98501 0.98679 0.98841 0.9898 0.99104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 PC17    PC18    PC19    PC20    PC21    PC22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.0011 0.00105 0.00098 0.00083 0.00075 0.00069</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.9921 0.99319 0.99417 0.99500 0.99575 0.99644</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 22  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 16 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DeconRNASeq...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> svd calculated PCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Importance of component(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  PC1    PC2     PC3     PC4     PC5    PC6     PC7     PC8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.6788 0.1242 0.08553 0.05229 0.01417 0.0082 0.00602 0.00494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.6788 0.8029 0.88847 0.94076 0.95493 0.9631 0.96915 0.97409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PC9    PC10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> R2            0.00393 0.00282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative R2 0.97802 0.98084</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Attention: the number of pure cell types = 10  defined in the signature matrix;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  PCA results indicate that the number of cell types in the mixtures = 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Epidish...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Preprocessing deconvolution features...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Checking consistency in deconvolution cell fractions across patients...............................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_BPRNACan_ is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_BPRNACanProMet is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_BPRNACan3DProMet is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_CBSX.HNSCC.scRNAseq is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_CBSX.Melanoma.scRNAseq is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_CBSX.NSCLC.PBMCs.scRNAseq is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_CCLE.TIL10 is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_TIL10 is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_LM22 is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_BPRNACan_ is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_BPRNACanProMet is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_BPRNACan3DProMet is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_CBSX.HNSCC.scRNAseq is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_CBSX.Melanoma.scRNAseq is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_CBSX.NSCLC.PBMCs.scRNAseq is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_CCLE.TIL10 is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_TIL10 is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination DeconRNASeq_LM22 is 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total sum across samples of combination Epidish_CBSX.Melanoma.scRNAseq is 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marcelo Hurtado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

