<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute deconvolution methods with variable signatures — compute_methods_variable_signature • multideconv</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute deconvolution methods with variable signatures — compute_methods_variable_signature"><meta name="description" content="Compute deconvolution methods with variable signatures"><meta property="og:description" content="Compute deconvolution methods with variable signatures"><meta property="og:image" content="https://mhurtado13.github.io/multideconv/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">multideconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/multideconv.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/VeraPancaldiLab/multideconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute deconvolution methods with variable signatures</h1>
      <small class="dont-index">Source: <a href="https://github.com/VeraPancaldiLab/multideconv/blob/main/R/cell_deconvolution.R" class="external-link"><code>R/cell_deconvolution.R</code></a></small>
      <div class="d-none name"><code>compute_methods_variable_signature.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compute deconvolution methods with variable signatures</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compute_methods_variable_signature</span><span class="op">(</span></span>
<span>  <span class="va">TPM_matrix</span>,</span>
<span>  <span class="va">signatures</span>,</span>
<span>  algos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CBSX"</span>, <span class="st">"Epidish"</span>, <span class="st">"DeconRNASeq"</span>, <span class="st">"DWLS"</span>, <span class="st">"MOMF"</span><span class="op">)</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">cbsx.name</span>,</span>
<span>  <span class="va">cbsx.token</span>,</span>
<span>  doParallel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  workers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sc_obj <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tpm-matrix">TPM_matrix<a class="anchor" aria-label="anchor" href="#arg-tpm-matrix"></a></dt>
<dd><p>A matrix with TPM normalized counts (samples as columns and genes symbols as rows)</p></dd>


<dt id="arg-signatures">signatures<a class="anchor" aria-label="anchor" href="#arg-signatures"></a></dt>
<dd><p>A path with a directory where signatures are located</p></dd>


<dt id="arg-algos">algos<a class="anchor" aria-label="anchor" href="#arg-algos"></a></dt>
<dd><p>A character vector with the methods to compute (Default methods are CBSX, Epidish, DeconRNASeq and DWLS)</p></dd>


<dt id="arg-exclude">exclude<a class="anchor" aria-label="anchor" href="#arg-exclude"></a></dt>
<dd><p>(Optional) A character vector with the signature to exclude</p></dd>


<dt id="arg-cbsx-name">cbsx.name<a class="anchor" aria-label="anchor" href="#arg-cbsx-name"></a></dt>
<dd><p>CIBERSORTx credential mail if CBSX will be run</p></dd>


<dt id="arg-cbsx-token">cbsx.token<a class="anchor" aria-label="anchor" href="#arg-cbsx-token"></a></dt>
<dd><p>CIBERSORTx credential token if CBSX will be run</p></dd>


<dt id="arg-doparallel">doParallel<a class="anchor" aria-label="anchor" href="#arg-doparallel"></a></dt>
<dd><p>Boolean value to specify if DWLS and CBSX should run in parallel (default is False)</p></dd>


<dt id="arg-workers">workers<a class="anchor" aria-label="anchor" href="#arg-workers"></a></dt>
<dd><p>Number of worker process to run during parallelization (default is NULL)</p></dd>


<dt id="arg-sc-obj">sc_obj<a class="anchor" aria-label="anchor" href="#arg-sc-obj"></a></dt>
<dd><p>A matrix with the counts from scRNAseq object (genes as rows and cells as columns) to run MOMF method. If NULL, MOMF is ignored.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix with the deconvolution features corresponding to all combinations of methods-signatures specified</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Sturm, G., Finotello, F., Petitprez, F., Zhang, J. D., Baumbach, J., Fridman, W. H., ..., List, M., Aneichyk, T. (2019). Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology.
Bioinformatics, 35(14), i436-i445. https://doi.org/10.1093/bioinformatics/btz363</p>
<p>Benchmarking second-generation methods for cell-type deconvolution of transcriptomic data. Dietrich, Alexander and Merotto, Lorenzo and Pelz, Konstantin and Eder, Bernhard and Zackl, Constantin and Reinisch, Katharina and
Edenhofer, Frank and Marini, Federico and Sturm, Gregor and List, Markus and Finotello, Francesca. (2024) https://doi.org/10.1101/2024.06.10.598226</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marcelo Hurtado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

