<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute second-generation deconvolution methods — compute_sc_deconvolution_methods • multideconv</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute second-generation deconvolution methods — compute_sc_deconvolution_methods"><meta name="description" content="Compute second-generation deconvolution methods"><meta property="og:description" content="Compute second-generation deconvolution methods"><meta property="og:image" content="https://mhurtado13.github.io/multideconv/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">multideconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/multideconv.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/VeraPancaldiLab/multideconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute second-generation deconvolution methods</h1>
      <small class="dont-index">Source: <a href="https://github.com/VeraPancaldiLab/multideconv/blob/main/R/cell_deconvolution.R" class="external-link"><code>R/cell_deconvolution.R</code></a></small>
      <div class="d-none name"><code>compute_sc_deconvolution_methods.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compute second-generation deconvolution methods</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compute_sc_deconvolution_methods</span><span class="op">(</span></span>
<span>  <span class="va">raw_counts</span>,</span>
<span>  normalized <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  methods_sc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Autogenes"</span>, <span class="st">"BayesPrism"</span>, <span class="st">"Bisque"</span>, <span class="st">"CPM"</span>, <span class="st">"MuSic"</span>, <span class="st">"SCDC"</span><span class="op">)</span>,</span>
<span>  <span class="va">sc_object</span>,</span>
<span>  <span class="va">sc_metadata</span>,</span>
<span>  <span class="va">cell_annotations</span>,</span>
<span>  <span class="va">samples_ids</span>,</span>
<span>  <span class="va">name_object</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  file_name <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-raw-counts">raw_counts<a class="anchor" aria-label="anchor" href="#arg-raw-counts"></a></dt>
<dd><p>A matrix with raw counts (samples as columns and genes symbols as rows)</p></dd>


<dt id="arg-normalized">normalized<a class="anchor" aria-label="anchor" href="#arg-normalized"></a></dt>
<dd><p>Boolean value to specify if raw_counts need to be normalized (If no raw_counts are available and argument corresponds to already normalized counts this arguments needs to be set to False)</p></dd>


<dt id="arg-methods-sc">methods_sc<a class="anchor" aria-label="anchor" href="#arg-methods-sc"></a></dt>
<dd><p>A character vector with the sc-deconvolution methods to run. Default are "Autogenes", "BayesPrism", "Bisque", "CPM", "MuSic", "SCDC"</p></dd>


<dt id="arg-sc-object">sc_object<a class="anchor" aria-label="anchor" href="#arg-sc-object"></a></dt>
<dd><p>A matrix with the counts from scRNAseq object (genes as rows and cells as columns)</p></dd>


<dt id="arg-sc-metadata">sc_metadata<a class="anchor" aria-label="anchor" href="#arg-sc-metadata"></a></dt>
<dd><p>Dataframe with metadata from the single cell object. The matrix should include the columns cell_label and sample_label.</p></dd>


<dt id="arg-cell-annotations">cell_annotations<a class="anchor" aria-label="anchor" href="#arg-cell-annotations"></a></dt>
<dd><p>A string with the column name with the cell labels (column should be of the same order as in the sc_object)</p></dd>


<dt id="arg-samples-ids">samples_ids<a class="anchor" aria-label="anchor" href="#arg-samples-ids"></a></dt>
<dd><p>A string with the column name with the samples labels (column should be of the same order as in the sc_object)</p></dd>


<dt id="arg-name-object">name_object<a class="anchor" aria-label="anchor" href="#arg-name-object"></a></dt>
<dd><p>Signature name to use in the generated single cell signature for deconvolving the bulk RNAseq data</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>Number of cores to use for paralellization. If no number is set, detectCores() - 1 will be set as the number.</p></dd>


<dt id="arg-return">return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>Whether to save or not the csv file with the deconvolution features.</p></dd>


<dt id="arg-file-name">file_name<a class="anchor" aria-label="anchor" href="#arg-file-name"></a></dt>
<dd><p>File name for the .csv file to save with the deconvolution results.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix of deconvolution features across samples from your bulk counts based on the second generation methods.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Sturm, G., Finotello, F., Petitprez, F., Zhang, J. D., Baumbach, J., Fridman, W. H., ..., List, M., Aneichyk, T. (2019). Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology.
Bioinformatics, 35(14), i436-i445. https://doi.org/10.1093/bioinformatics/btz363</p>
<p>Benchmarking second-generation methods for cell-type deconvolution of transcriptomic data. Dietrich, Alexander and Merotto, Lorenzo and Pelz, Konstantin and Eder, Bernhard and Zackl, Constantin and Reinisch, Katharina and
Edenhofer, Frank and Marini, Federico and Sturm, Gregor and List, Markus and Finotello, Francesca. (2024) https://doi.org/10.1101/2024.06.10.598226</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marcelo Hurtado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

