<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create cell type signatures from scRNAseq — create_sc_signatures • multideconv</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create cell type signatures from scRNAseq — create_sc_signatures"><meta name="description" content="Create cell type signatures from scRNAseq"><meta property="og:description" content="Create cell type signatures from scRNAseq"><meta property="og:image" content="https://mhurtado13.github.io/multideconv/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">multideconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/multideconv.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/VeraPancaldiLab/multideconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Create cell type signatures from scRNAseq</h1>
      <small class="dont-index">Source: <a href="https://github.com/VeraPancaldiLab/multideconv/blob/main/R/cell_deconvolution.R" class="external-link"><code>R/cell_deconvolution.R</code></a></small>
      <div class="d-none name"><code>create_sc_signatures.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Create cell type signatures from scRNAseq</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_sc_signatures</span><span class="op">(</span></span>
<span>  <span class="va">sc_obj</span>,</span>
<span>  <span class="va">sc_metadata</span>,</span>
<span>  <span class="va">cells_labels</span>,</span>
<span>  <span class="va">sample_labels</span>,</span>
<span>  credentials.mail <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  credentials.token <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bulk_rna <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name_signature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  methods_sig <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DWLS"</span>, <span class="st">"CIBERSORTx"</span>, <span class="st">"MOMF"</span>, <span class="st">"BSeqsc"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sc-obj">sc_obj<a class="anchor" aria-label="anchor" href="#arg-sc-obj"></a></dt>
<dd><p>A matrix with the counts from scRNAseq object (genes as rows and cells as columns)</p></dd>


<dt id="arg-sc-metadata">sc_metadata<a class="anchor" aria-label="anchor" href="#arg-sc-metadata"></a></dt>
<dd><p>Dataframe with metadata from the single cell object. The matrix should include the columns cell_label and sample_label.</p></dd>


<dt id="arg-cells-labels">cells_labels<a class="anchor" aria-label="anchor" href="#arg-cells-labels"></a></dt>
<dd><p>A character vector with the cell labels (need to be of the same order as in the sc_object)</p></dd>


<dt id="arg-sample-labels">sample_labels<a class="anchor" aria-label="anchor" href="#arg-sample-labels"></a></dt>
<dd><p>A character vector with the samples labels (need to be of the same order as in the sc_object)</p></dd>


<dt id="arg-credentials-mail">credentials.mail<a class="anchor" aria-label="anchor" href="#arg-credentials-mail"></a></dt>
<dd><p>(Optional) Credential email for running CIBERSORTx If not provided, CIBERSORTx method will not be run.</p></dd>


<dt id="arg-credentials-token">credentials.token<a class="anchor" aria-label="anchor" href="#arg-credentials-token"></a></dt>
<dd><p>(Optional) Credential token for running CIBERSORTx. If not provided, CIBERSORTx method will not be run.</p></dd>


<dt id="arg-bulk-rna">bulk_rna<a class="anchor" aria-label="anchor" href="#arg-bulk-rna"></a></dt>
<dd><p>A matrix of bulk data. Rows are genes, columns are samples. This is needed for MOMF method, if not given the method will not be run.</p></dd>


<dt id="arg-cell-markers">cell_markers<a class="anchor" aria-label="anchor" href="#arg-cell-markers"></a></dt>
<dd><p>Named list with the genes markers names as Symbol per cell types to be used to create the signature using the BSeq-SC method. If NULL, the method will be ignored during the signature creation.</p></dd>


<dt id="arg-name-signature">name_signature<a class="anchor" aria-label="anchor" href="#arg-name-signature"></a></dt>
<dd><p>A string indicating the signature name. This will be added as a suffix in each method (e.g. CBSX_name_signature, DWLS_name_signature)</p></dd>


<dt id="arg-methods-sig">methods_sig<a class="anchor" aria-label="anchor" href="#arg-methods-sig"></a></dt>
<dd><p>A character vector specifying which methods to run. Options are "DWLS", "CIBERSORTx", "MOMF", and "BSeqsc". Default runs all available methods.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the cell signatures per method. Signatures are directly saved in Results/custom_signatures folder, these will be used to run deconvolution.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Sturm, G., Finotello, F., Petitprez, F., Zhang, J. D., Baumbach, J., Fridman, W. H., ..., List, M., Aneichyk, T. (2019). Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology.
Bioinformatics, 35(14), i436-i445. https://doi.org/10.1093/bioinformatics/btz363</p>
<p>Benchmarking second-generation methods for cell-type deconvolution of transcriptomic data. Dietrich, Alexander and Merotto, Lorenzo and Pelz, Konstantin and Eder, Bernhard and Zackl, Constantin and Reinisch, Katharina and
Edenhofer, Frank and Marini, Federico and Sturm, Gregor and List, Markus and Finotello, Francesca. (2024) https://doi.org/10.1101/2024.06.10.598226</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marcelo Hurtado.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

